source/develop/release-management/features/virt/vm-screenshot.md

```diff
@@ -0,0 +1,83 @@
+---
+title: VM Screenshot
+category: feature
+authors: sabusale
+---
+
+# VM Screenshot
+
+## Summary
+
+Allow users to take screenshot of the VM
+
+## Owner
+
+*   Name: Saif Abu Saleh (sabusale)
+*   Email: sabusale@redhat.com
+
+## Description
+
+User can take a screenshot of running VM
+
+### UI
+
+Not supported
+
+### Backend
+
+Backend will call VDSM to capture screenshot and will receive the `ppm`[[1]] file content
+and then `ppm` file content will be returned
+
+### VDSM
+
+VDSM will call Libvirt API[[2]] that capture the screenshot and saves it on given path on the host file system
+and then will return the `ppm` file content and remove the image on the host file system
+
+Saving the image:
+*   path will be temporary directory path
```

Comment:
I think we have a vm-specific directory we can place it in

Reply:
Both libvirt (under /var/lib/libvirt/qemu) and Vdsm (under /run/vdsm) create VM specific subdirectories, not necessarily for general use though.

Reply:
looks like definitely /var/lib/libvirt/ could be good place (Like TPM) or /var/lib/libvirt/qemu/

but again, depending what we want to do with this

If we don't want to persist the images in the host, I think the directory path shouldn't really matter and creating it in random place is enough, since it will be deleted

otherwise, if we want to keep the images in the host, then its really matter

Reply:
IIUC, we don't have an alternative but to persist it on the host and then read it and send it to the engine, right?
I think that by using an existing per-VM folder we are (1) more consistent (as you wrote, e.g., with TPM); and (2) assuming we already clear those folders when destroying VMs, it would simplify handling of negative flows

Reply:
agree

the thing is that we need such a folder, such that it will get deleted when the VM is destroyed

for tpm and nvram from what I checked this folders is created and managed by Libvirt

this is not the case here, since not specifying path will generate the screenshot on the root directory

Reply:
Using a /var/lib/libvirt subdirectory is probably not a good idea, this directory belongs to libvirt and we shouldn't mess with it. As for TPM and NVRAM, we use it only to read and restore data managed by libvirt, as already pointed out by Saif.
Using a /run subdirectory is probably also not a good idea, this directory servers AFAIK for files describing the state of the system, such as pid files, mount points, etc., not for data dumps.
/tmp or a similar location looks like the best place for such a temporary data that is needed only during the API call. Using Python's tempfile.TemporaryDirectory should do the right thing, including selecting the right and safe location.

Reply:
@mz-pdm ok, if we do that in a single call, i.e., call libvirt to dump the data and then immediately read the data, send it to the engine and remove the data, then yeah, a temporary folder sounds ok. but if we are implementing it in a way to prepares it for asynchronous operation, I'd prefer to avoid sending the path we dump the data back to the engine so the engine would later provide it back to VDSM or to cache it in VDSM, I think in that case it would be better to rely on some convention for where the data is stored

Reply:
Then a subdirectory of /var/lib/vdsm could be used for such a bit more persistent purpose.

Reply:
few questions:
1- in /var/lib/vdsm Can we just do something like: /var/lib/vdsm/screenshots and then the screenshots names contain the VM name on them (see page Saving the image section)

2- what will happen when upgrading VDSM host to this directory?

Reply:
> 1- in /var/lib/vdsm Can we just do something like: /var/lib/vdsm/screenshots and then the screenshots names contain the VM name on them (see page Saving the image section)

Yes. There just needs to be some form of request-file mapping in case of asynchronous requests. And maybe using VM ids rather than names in the file names would be better -- less user friendly but more robust.

> 2- what will happen when upgrading VDSM host to this directory?

In the default/recommended setup, Vdsm cannot be upgraded when VMs are running on the host. The directory can simply stay there. Some content cleanup mechanism should be considered; maybe using systemd-tmpfiles would be the easiest way.

Reply:
we can say we save to the static path  (e.g /var/lib/vdsm/screenshots) with file-name as generated in libvirt (<vm_name>-<timestamp>) for now

And the other parts (async request and all its requested work) will be in Future work